### PR TITLE
[capture] Fixes for SHA3 capture

### PR DESCRIPTION
**Pull request by vogelpi**
This PR contains two commits for the SHA3 capture. The second one is very relevant, I think this might potentially resolve also the issue @y-srini has been facing. I've found this out using the plots added with PR #121 .

### CODE REVIEW

**Comment by vogelpi** [comment]
FYI: I am not yet entirely sure why only observed this fake leakage with new bitstreams and not the one in the repo currently. I created multiple new bitstreams and they all showed the same behavior but to a different extent. It could be due to P&R variation. I observed that over the last couple of months LUT utilization increased again to almost 85% with masked KMAC . But there were no RTL changes in KMAC and relevant prims which could have caused this.

Also with the new fixed plaintext the behavior occurs but it's much weakger and becomes visible at higher number of traces.

File: cw/capture.py [hunk 1]
```diff
@@ -593,8 +593,27 @@ def capture_sha3_fvsr_data_batch(ot, ktp, capture_cfg, scope_type):
...
+    # levels for the fixed and random trace set, potentially leading to a small vertical offset
+    # between the two trace sets. This offset is detectable by TVLA and covers actual leakage
+    # happening during the SHA3 processing. The effect is most easliy visible between loading the
+    # plaintext and appending the padding, i.e., when the target is completely idle and waiting for
+    # the 40 clock cycle timer delay between the RUN and PROCESS command to expire.
+    #
+    # Crafted plaintext value with 4 bits set per byte, and where the Hamming distance between the
+    # first and second 64-bit word is exatly 4 bits per byte.
+    plaintext_fixed = bytearray([0xA5, 0xC3, 0x5A, 0x3C, 0x96, 0x0F, 0x69, 0xF0,
+                                 0xC3, 0xA5, 0x3C, 0x5A, 0x0F, 0x96, 0xF0, 0x69])
```
Comment (vrozic) [inline cw/capture.py:616]: My preference would be to hide this change behind a parameter loaded from the cfg_file. Something like `capture_cfg["ballanced_fixed_ptext"]`.
This is essentially a debugging feature and we would eventually want to solve the problem in a more robust way.

**Review by vrozic — APPROVED** [review]

File: cw/capture.py (hunk 1 above)
Comment (vogelpi) [reply cw/capture.py:616]: That's a good point @vrozic , I'll try to do something like this. Thanks for your review.

**Comment by y-srini** [comment]
@vogelpi, Thanks for the update. 

Ran TVLA with the balanced plaintexts, on Mask OFF and Mask ON setting. 
Picture below compares the TVLA without and with balanced plaintext (using same bitstreams with which the shift was observed). 

Balanced plaintext definitely seem improve things here, to reduce fake leakages. 

image

**Comment by abdullahvarici** [comment]
I think balancing the plaintext is a good point. Maybe we should also think about this: There is always a possibility of capturing two or more traces for fixed key consecutively since we are determining randomly on runtime whether we are going to send a fixed key input or random. Maybe we should prevent it because some values may not be change if we run the algorithm for the same key causing zero Hamming Weight. But maybe there is already a mechanism like secure wipe. Just an idea.

**Review by abdullahvarici — COMMENTED** [review]
LGTM!

**Comment by vogelpi** [comment]
FYI: @johannheyszl and I were discussing further about this. After some thinking we thought it would be really interesting that the vertical offset is just introduced by the plaintext loading and not e.g. by previous processing on Ibex or TL-UL transfers (high-fanout nets).

However, it seems that really what is happening. The figure below shows the difference between average traces in the fvsr data test (100k traces) around the plaintext loading. Any offset due to Ibex processing / TL-UL transfers would be to the left of the plaintext loading. As you can see, the difference is nicely centered around 0 there. The offset is only introduced after the peak and then slowly decays back to 0.

Around 1000 samples (50 clock cycles) after the plaintext loading, things are back to normal. However, the time it takes for things to normalize heavily depends on the actual bitstream.

sha3_plaintext_loading

**Pull request merged**